File: algorithm/c_algorithm/src/a.1-3_circular_linked_lista/CircularLinkedList.c
/*
 * CirtularLinkedList.c
 *
 *  Created on: 2018. 12. 30.
 *      Author: otrodevym
 */

#include "CircularLinkedList.h"

Node* CDLL_CreateNode(ElementType newData) {
	Node* NewNode = (Node*) malloc(sizeof(Node));

	NewNode->Data = newData;
	NewNode->PrevNode = NULL;
	NewNode->NextNode = NULL;
	return NewNode;
}

void CDLL_DestroyNode(Node* Node) {
	free(Node);
}

void CDLL_AppendNode(Node** Head, Node* NewNode) {
	if ((*Head) == NULL) {
		*Head = NewNode;
		(*Head)->NextNode = *Head;
		(*Head)->PrevNode = *Head;
	} else {
		Node *Tail = (*Head)->PrevNode;

		Tail->NextNode->PrevNode = NewNode;
		Tail->NextNode = NewNode;

		NewNode->NextNode = (*Head);
		NewNode->PrevNode = Tail;
	}
}

void CDLL_InsertAfter(Node* Current, Node* NewNode) {
	NewNode->NextNode = Current->NextNode;
	NewNode->PrevNode = Current;

	Current->NextNode->PrevNode = NewNode;
	Current->NextNode = NewNode;
}


void CDLL_RemoveNode(Node** Head, Node* Remove){
	if( *Head == NULL){
		(*Head)->PrevNode->NextNode = Remove->NextNode;
		(*Head)->NextNode->PrevNode = Remove->PrevNode;

		*Head = Remove->NextNode;

		Remove->PrevNode = NULL;
		Remove->NextNode = NULL;
	}else{
		Node* Temp = Remove;
		Remove->PrevNode->NextNode = Temp->NextNode;
		Remove->NextNode->PrevNode = Temp->PrevNode;

		Remove->PrevNode = NULL;
		Remove->NextNode = NULL;
	}
}

Node* CDLL_GetNodeAt(Node* Head, int Location){
	Node* Current = Head;
	while(Current != NULL && (--Location) >=0){
		Current = Current->NextNode;
	}
	return Current;
}

int CDLL_GetNodeCount(Node* Head){
	unsigned int Count = 0;
	Node* Current = Head;
	while(Current != NULL ){
		Current = Current->NextNode;
		Count++;
		if(Current == Head){
			break;
		}
	}
	return Count;
}


void PrintNode(Node* _Node){
	if(_Node->PrevNode == NULL){
		printf("prev : NULL");
	}else{
		printf("prev : %d ", _Node->PrevNode->Data);
	}
	printf(" Current : %d ", _Node->Data);

	if(_Node->NextNode == NULL){
		printf("Next : NULL \n");
	}else{
		printf("Next : %d \n", _Node->NextNode->Data);
	}
}

File: algorithm/[book]Java_Algorithm/src/a_linked_list/CircularLinkedList.java
package a_linked_list;

public class CircularLinkedList<T> {
	private int size;
	private Node head;
	
	class Node{
		private T value;
		private Node prevNode;
		private Node nextNode;
		
		public Node(T value) {
			this.value = value;
			this.prevNode = null;
			this.nextNode = null;
		}
	}
	public CircularLinkedList() {
		this.size = 0;
		this.head = null;
	}
	
	
	private void add(T value) {
		Node newNode = new Node(value);
		if(this.head == null) {
			this.head = newNode;
			this.head.nextNode = newNode;
			this.head.prevNode = newNode;
		}else {
			
			Node current = this.head;
			
			current.prevNode.nextNode = newNode;
			newNode.prevNode = current.prevNode;
			
			newNode.nextNode = this.head;
			this.head.prevNode= newNode;
		}
		this.size++;
	}
	
	private void remove(int index) {
		Node current = this.head;
		while(current.nextNode != null && (--index) >=0) {
			current = current.nextNode;
		}
		current.nextNode.prevNode = current.prevNode;
		current.prevNode.nextNode = current.nextNode;
		current.nextNode = null;
		current.prevNode = null;
		
		this.size--;
	}
	
	private T get(int index) {
		Node current = this.head;
		if(0 == index) {
			
//			환형 링크드 리스트는 계속 돌아야 맞음
//		}else if(index == (size-1)) {
//			while(current.nextNode != null && (--index) >= 0) {
//				current = current.nextNode;
//			}
		}else {
			while(current.nextNode != null && (--index) >= 0) {
				current = current.nextNode;
			}
		}
		return current.value;
	}
	
	private void update(int index, T value) {
		Node current = this.head;
		while(current.nextNode != null && (--index) >= 0){
			current = current.nextNode;
		}
		
		current.value = value;
		
	}
	
	private void print() {
		for(int i = 0; i < size; i++) {
			System.out.println(this.get(i));
		}
	}
	
	
	public static void main(String args[]) {
		CircularLinkedList<String> list = new CircularLinkedList<>();
		list.add("1");
		list.add("2");
		list.add("3");
		list.add("4");
		list.print();
		System.out.println("----------");
		
		list.remove(1);
		list.print();
		System.out.println("----------");
		
		list.update(1, "100");
		list.print();
		System.out.println("----------");
		
		list.update(3, "100");
		list.print();
		System.out.println("----------");
		
		
	}
}
